File: ops/test_crm_collaborator_profile.mjs
import assert from "node:assert/strict";
import http from "node:http";
import { existsSync } from "node:fs";
import { readFile } from "node:fs/promises";
import { chromium } from "playwright";

const bridge = await readFile(new URL("../app/public/crm-collaborator-profile.js", import.meta.url), "utf8");
let achievements = [];
const profilePayload = () => ({
  profile: {
    id: 7, name: "Matheus Henrique", email: "matheuscrc@instituto.local",
    crm_access_level: "admin", service_sector: "CRC", photo_url: null,
    stats: {active_count:0,resolved_month:14,attendances_month:22,first_consultations:8,recoveries:5}
  },
  achievements,
  can_manage: true,
  collaborators: [{id:7,name:"Matheus Henrique",crm_access_level:"admin"},{id:8,name:"Isabela Cristina",crm_access_level:"attendant"}]
});

const pageHtml = `<!doctype html><html><body data-omtheme="light" style="margin:0;background:var(--bg,#f0f2f5)">
  <aside style="width:80px;height:100vh;display:flex;flex-direction:column;align-items:center;background:#122b46">
    <div style="flex:1"></div>
    <div id="theme" style="width:44px;height:38px;cursor:pointer"><svg viewBox="0 0 24 24"><path d="M21 12.79A9 9 0 1 1 11.21 3"></path></svg></div>
    <div id="user-menu" style="position:relative;cursor:pointer"><div id="avatar" style="width:42px;height:42px;border-radius:50%;cursor:pointer">AS</div><span style="position:absolute;width:13px;height:13px"></span></div>
  </aside><script>
    document.getElementById("theme").addEventListener("click", function () {
      document.body.dataset.omtheme = document.body.dataset.omtheme === "dark" ? "light" : "dark";
    });
    document.getElementById("user-menu").addEventListener("click", function () {
      document.body.dataset.legacyStatusMenuOpened = "true";
    });
  </script><script src="/crm-collaborator-profile.js"></script></body></html>`;

const server = http.createServer(async (request, response) => {
  if (request.url === "/crm-collaborator-profile.js") {
    response.writeHead(200, {"Content-Type":"text/javascript; charset=utf-8"});
    return response.end(bridge);
  }
  if (request.url === "/api/auth/status") {
    response.writeHead(200, {"Content-Type":"application/json"});
    return response.end(JSON.stringify({authenticated:true,user:{id:7,name:"Matheus Henrique",role:"crc",crm_access_level:"admin"}}));
  }
  if (request.url.startsWith("/api/crm/profile/achievements") && request.method === "POST") {
    let raw = ""; for await (const chunk of request) raw += chunk;
    const item = JSON.parse(raw);
    achievements = [{id:1,source:"manual",title:item.title,description:item.description,icon_key:item.icon_key,accent_color:item.accent_color,awarded_at:"2026-08-02 15:00:00",awarded_by:"Matheus Henrique"}];
    response.writeHead(201, {"Content-Type":"application/json"});
    return response.end(JSON.stringify({created:true,achievement:achievements[0]}));
  }
  if (request.url.startsWith("/api/crm/profile")) {
    response.writeHead(200, {"Content-Type":"application/json"});
    return response.end(JSON.stringify(profilePayload()));
  }
  response.writeHead(200, {"Content-Type":"text/html; charset=utf-8"});
  response.end(pageHtml);
});

await new Promise(resolve => server.listen(0, "127.0.0.1", resolve));
const executablePath = [process.env.PLAYWRIGHT_CHROMIUM_EXECUTABLE_PATH,"C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe","C:\\Program Files (x86)\\Microsoft\\Edge\\Application\\msedge.exe"].filter(Boolean).find(existsSync);
const browser = await chromium.launch({headless:true,...(executablePath ? {executablePath} : {})});
try {
  const page = await browser.newPage({viewport:{width:1440,height:900}});
  await page.goto(`http://127.0.0.1:${server.address().port}/`);
  const avatar = page.getByRole("button", {name:"Abrir meu perfil e conquistas"});
  await avatar.waitFor();
  assert.equal(await avatar.textContent(), "MH", "o avatar legado AS deve receber as iniciais reais");
  const theme = page.getByRole("button", {name:"Alternar modo noturno"});
  await theme.waitFor(); await theme.click();
  await page.waitForFunction(() => document.body.dataset.omtheme === "dark");
  assert.equal(await page.evaluate(() => localStorage.getItem("iea.crm.theme")), "dark");

  await avatar.evaluate(element => {
    const replacement = element.cloneNode(true);
    element.replaceWith(replacement);
    replacement.click();
  });
  await page.getByRole("heading", {name:"Matheus Henrique"}).waitFor();
  assert.equal(await page.evaluate(() => document.body.dataset.legacyStatusMenuOpened || "false"), "false");
  assert.equal(await page.getByRole("dialog").count(), 1);
  assert.equal(await page.getByText("Administrador do CRM", {exact:true}).count(), 1);
  assert.equal(await page.getByText("22", {exact:true}).count(), 1);
  assert.equal(await page.getByText("Este perfil ainda não possui conquistas.").count(), 1);

  await page.getByRole("button", {name:"+ Criar conquista"}).click();
  await page.getByLabel("Título da conquista").fill("Excelência no atendimento");
  await page.getByLabel("Mensagem de reconhecimento").fill("Destaque pelo cuidado com os pacientes.");
  await page.getByLabel("Símbolo").selectOption("medal");
  await page.getByRole("button", {name:"Publicar conquista"}).click();
  await page.getByRole("heading", {name:"Excelência no atendimento"}).waitFor();
  assert.equal(await page.getByText("Destaque criado por Matheus Henrique", {exact:false}).count(), 1);

  await page.setViewportSize({width:390,height:844});
  const box = await page.getByRole("dialog").boundingBox();
  assert.equal(Math.round(box.width), 390);
  assert.equal(Math.round(box.height), 844);
  console.log("crm-collaborator-profile-e2e-ok");
} finally {
  await browser.close();
  await new Promise(resolve => server.close(resolve));
}

File: ops/test_crm_collaborator_profile.py
"""Regression checks for CRM collaborator profiles and manual achievements."""

import sqlite3
import sys
import types
from http import HTTPStatus
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT / "app"))

if "qrcode" not in sys.modules:
    qrcode = types.ModuleType("qrcode")
    qrcode_image = types.ModuleType("qrcode.image")
    qrcode_svg = types.ModuleType("qrcode.image.svg")
    qrcode_svg.SvgPathImage = object
    qrcode.image = qrcode_image
    sys.modules.update({"qrcode": qrcode, "qrcode.image": qrcode_image, "qrcode.image.svg": qrcode_svg})
if "openpyxl" not in sys.modules:
    openpyxl = types.ModuleType("openpyxl")
    openpyxl.load_workbook = lambda *args, **kwargs: None
    sys.modules["openpyxl"] = openpyxl
if "psycopg" not in sys.modules:
    psycopg = types.ModuleType("psycopg")
    psycopg_errors = types.ModuleType("psycopg.errors")
    class PsycopgError(Exception):
        pass
    class IntegrityError(PsycopgError):
        pass
    psycopg.Error = PsycopgError
    psycopg.errors = psycopg_errors
    psycopg.sql = types.SimpleNamespace()
    psycopg_errors.IntegrityError = IntegrityError
    sys.modules.update({"psycopg": psycopg, "psycopg.errors": psycopg_errors})

import server  # noqa: E402


def test_schema_and_frontend_contract() -> None:
    schema = (ROOT / "app" / "schema.sql").read_text(encoding="utf-8")
    db = sqlite3.connect(":memory:")
    db.executescript(schema)
    columns = {row[1] for row in db.execute("PRAGMA table_info(crm_profile_achievements)")}
    assert {"user_id", "title", "description", "icon_key", "accent_color", "awarded_by_user_id", "active"} <= columns

    bridge = (ROOT / "app" / "public" / "crm-collaborator-profile.js").read_text(encoding="utf-8")
    html = (ROOT / "app" / "public" / "crm-whatsapp.html").read_text(encoding="utf-8")
    assert 'data-iea-cp-trigger' in bridge
    assert 'iea.crm.theme' in bridge and 'data-omtheme' in bridge
    assert 'prefers-reduced-motion:reduce' in bridge
    assert '/api/crm/profile/achievements' in bridge
    assert 'role="dialog"' in bridge and 'aria-modal="true"' in bridge
    assert 'crm-collaborator-profile.js?v=20260803-profile-trophies-v1' in html
    assert 'bundle legado ainda desenha o rodapé com as iniciais fixas "AS"' in bridge
    assert 'document.addEventListener("click", interceptProfileTrigger, true)' in bridge


def test_only_crm_admin_can_create_manual_achievement() -> None:
    db = sqlite3.connect(":memory:")
    db.row_factory = sqlite3.Row
    db.executescript("""
        CREATE TABLE users(id INTEGER PRIMARY KEY,name TEXT,access_role TEXT,active INTEGER);
        CREATE TABLE crm_profile_achievements(
          id INTEGER PRIMARY KEY AUTOINCREMENT,user_id INTEGER,title TEXT,description TEXT,
          icon_key TEXT,accent_color TEXT,awarded_by_user_id INTEGER,
          awarded_at TEXT DEFAULT CURRENT_TIMESTAMP,active INTEGER DEFAULT 1
        );
        CREATE TABLE security_events(
          id INTEGER PRIMARY KEY AUTOINCREMENT,user_id INTEGER,event_type TEXT,
          detail TEXT,ip_address TEXT,created_at TEXT DEFAULT CURRENT_TIMESTAMP
        );
        INSERT INTO users VALUES(1,'Gestor','crc',1),(2,'Atendente','crc',1);
    """)
    original_connect = server.connect
    server.connect = lambda: db
    try:
        handler = server.ClinicHandler.__new__(server.ClinicHandler)
        sent = []
        handler.send_json = lambda payload, status=HTTPStatus.OK: sent.append((payload, status))
        handler.request_ip = lambda: "127.0.0.1"
        handler.authenticated_user = {
            "id": 2, "name": "Atendente", "access_role": "crc",
            "crm_access_level": "attendant", "permissions_json": "{}",
        }
        handler.create_crm_profile_achievement({"user_id": 2, "title": "Excelência"})
        assert sent[-1][1] == HTTPStatus.FORBIDDEN
        assert db.execute("SELECT COUNT(*) FROM crm_profile_achievements").fetchone()[0] == 0

        handler.authenticated_user = {
            "id": 1, "name": "Gestor", "access_role": "crc",
            "crm_access_level": "admin", "permissions_json": "{}",
        }
        handler.create_crm_profile_achievement({
            "user_id": 2, "title": "Excelência no atendimento",
            "description": "Reconhecimento pelo cuidado com os pacientes.",
            "icon_key": "medal", "accent_color": "#7C3AED",
        })
        assert sent[-1][1] == HTTPStatus.CREATED
        row = db.execute("SELECT * FROM crm_profile_achievements").fetchone()
        assert row["user_id"] == 2 and row["awarded_by_user_id"] == 1
        assert row["title"] == "Excelência no atendimento"
        assert db.execute("SELECT event_type FROM security_events").fetchone()[0] == "crm_profile_achievement_created"
    finally:
        server.connect = original_connect


def test_profile_visibility_is_limited_to_self_or_crm_admin() -> None:
    db = sqlite3.connect(":memory:")
    db.row_factory = sqlite3.Row
    db.executescript((ROOT / "app" / "schema.sql").read_text(encoding="utf-8"))
    db.executemany(
        "INSERT INTO users(id,name,email,access_role,crm_access_level,service_sector,active) VALUES(?,?,?,?,?,?,1)",
        (
            (1, "Gestor", "gestor@example.test", "crc", "admin", "CRC"),
            (2, "Atendente", "atendente@example.test", "crc", "attendant", "CRC"),
        ),
    )
    original_connect = server.connect
    server.connect = lambda: db
    try:
        handler = server.ClinicHandler.__new__(server.ClinicHandler)
        sent = []
        handler.send_json = lambda payload, status=HTTPStatus.OK: sent.append((payload, status))
        handler.authenticated_user = {
            "id": 2, "name": "Atendente", "access_role": "crc",
            "crm_access_level": "attendant", "permissions_json": "{}",
        }
        handler.get_crm_collaborator_profile({"user_id": ["2"]})
        assert sent[-1][1] == HTTPStatus.OK
        assert sent[-1][0]["profile"]["name"] == "Atendente"
        assert sent[-1][0]["can_manage"] is False
        assert sent[-1][0]["collaborators"] == []

        handler.get_crm_collaborator_profile({"user_id": ["1"]})
        assert sent[-1][1] == HTTPStatus.FORBIDDEN

        handler.authenticated_user = {
            "id": 1, "name": "Gestor", "access_role": "crc",
            "crm_access_level": "admin", "permissions_json": "{}",
        }
        handler.get_crm_collaborator_profile({"user_id": ["2"]})
        assert sent[-1][1] == HTTPStatus.OK
        assert sent[-1][0]["profile"]["name"] == "Atendente"
        assert sent[-1][0]["can_manage"] is True
        assert len(sent[-1][0]["collaborators"]) == 2
    finally:
        server.connect = original_connect


if __name__ == "__main__":
    test_schema_and_frontend_contract()
    test_only_crm_admin_can_create_manual_achievement()
    test_profile_visibility_is_limited_to_self_or_crm_admin()
    print("crm-collaborator-profile-tests-ok")
